airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py

```diff
@@ -548,3 +576,129 @@ def upload_metadata_to_gcs(bucket_name: str, metadata_file_path: Path, validator
+        )
+    )
+    version_blob.delete()
+    deleted_files.append(
+        DeletedFile(
+            id="version_metadata",
+            deleted=True,
+            description="versioned metadata",
+            blob_id=version_blob.id,
+        )
+    )
+
+    return MetadataDeleteInfo(
+        metadata_deleted=True,
+        deleted_files=deleted_files,
+    )
+
+
+def promote_release_candidate_in_gcs(
+    bucket_name: str, docker_repository: str, connector_version: str
+) -> Tuple[MetadataUploadInfo, MetadataDeleteInfo]:
+    """Promote a release candidate to the latest version in a GCS bucket.
+    The release candidate metadata file will be copied to the latest metadata file and then deleted.
+    We first check that the release candidate metadata file hash matches the version metadata file hash.
+    Args:
+        bucket_name (str): Name of the GCS bucket from which the release candidate will be deleted.
+        docker_repository (str): Name of the connector docker image.
+        connector_version (str): Version of the connector.
+    Returns:
+        Tuple[MetadataUploadInfo, MetadataDeleteInfo]: Information about the files that were uploaded (new latest version) and deleted (release candidate).
+    """
+
+    storage_client = _get_storage_client()
+    bucket = storage_client.bucket(bucket_name)
+
+    gcp_connector_dir = f"{METADATA_FOLDER}/{docker_repository}"
+    version_path = f"{gcp_connector_dir}/{connector_version}/{METADATA_FILE_NAME}"
+    rc_path = f"{gcp_connector_dir}/{RELEASE_CANDIDATE_GCS_FOLDER_NAME}/{METADATA_FILE_NAME}"
+    latest_path = f"{gcp_connector_dir}/{LATEST_GCS_FOLDER_NAME}/{METADATA_FILE_NAME}"
+
+    version_blob = bucket.blob(version_path)
+    latest_blob = bucket.blob(latest_path)
+    rc_blob = bucket.blob(rc_path)
+
+    if not version_blob.exists():
+        raise FileNotFoundError(f"Version metadata file {version_path} does not exist in the bucket.")
+    if not rc_blob.exists():
+        raise FileNotFoundError(f"Release candidate metadata file {rc_path} does not exist in the bucket.")
+
+    if rc_blob.md5_hash != version_blob.md5_hash:
+        raise ValueError(
+            f"""Release candidate metadata file {rc_path} hash does not match the version metadata file {version_path} hash. Unsafe to promote.
+            It's likely that something changed the release candidate hash but have not changed the metadata for the lastest matching version."""
+        )
+
+    uploaded_files = []
+    deleted_files = []
+
+    bucket.copy_blob(rc_blob, bucket, latest_blob)
+    uploaded_files.append(
+        UploadedFile(
+            id="latest_metadata",
+            uploaded=True,
+            blob_id=latest_blob.id,
+        )
+    )
+
+    rc_blob.delete()
+    deleted_files.append(
+        DeletedFile(
+            id="release_candidate_metadata",
+            deleted=True,
+            description="release candidate metadata",
+            blob_id=rc_blob.id,
+        )
+    )
```

Comment:
Something that can wait until @alafanechere is back -

I have a slight worry that we could end up in a situation where we've promoted the RC to latest but there's a failure before we can delete the RC. I wonder if we should actually do this cleanup in a separate task that can check to see if an RC's version is now in a registry of "normal" versions (i.e. all versions ever released to prod instead of just `latest` versions).

Reply:
I'll get this merged and try to test and see what blows up. Wouldn't be surprised if we'd need to clean it up a bit.